R/kenwardroger.R

```diff
@@ -0,0 +1,160 @@
+#' @param object `mmrm` object
+#' @param contrast contrast vector
+#' @param linear `logical`\cr whether to use linear Kenward-Roger approximation
+#' @return List with `est`, `se`, `df`, `t_stat` and `p_val`.
+#' @keywords internal
+df_1d_kr <- function(object, contrast, linear = FALSE) {
+  if (object$tmb_data$reml != 1) {
+    stop("Kenward-Roger is only for REML!")
+  }
+  assert_numeric(contrast, len = length(component(object, "beta_est")))
+  est <- sum(contrast * component(object, "beta_est"))
+  kr_comp <- object$kr_comp
+  w <- solve(object$tmb_obj$he(object$theta_est))
+  v_adj <- object$beta_vcov_adj
+  df <- h_kr_df(object$beta_vcov, matrix(contrast, nrow = 1), w, kr_comp$P)
+  se <- sqrt(contrast %*% v_adj %*% contrast)[1, 1]
+
+  ret <- list(
+    est = est,
+    se = se,
+    df = df$m
+  )
+  t_stat <- abs(est) / se
+  ret$t_stat <- t_stat
+  ret$p_val <- pt(t_stat, df$m, lower.tail = FALSE) / 2
+  ret
+}
+
+#' obtain the adjusted Kenward-Roger degree of freedom
+#' @param v0 unadjusted covariance matrix
+#' @param l linear combination matrix
+#' @param w hessian matrix
+#' @param p P matrix from `h_get_kr_comp`
+#' @return List with `m` and `lambda`
+#' @keywords internal
+h_kr_df <- function(v0, l, w, p) {
+  theta <- t(l) %*% solve(l %*% v0 %*% t(l)) %*% l
+  nl <- nrow(l)
+  thetav0 <- theta %*% v0
+  pl <- lapply(seq_len(nrow(p) / ncol(p)), function(x) {
+    ii <- (x - 1) * ncol(p) + 1
+    jj <- x * ncol(p)
+    p[ii:jj, ]
+  })
+  thetav0pv0 <- lapply(pl, function(x) {
+    thetav0 %*% x %*% v0
+  })
+  a1 <- 0
+  a2 <- 0
+  for (i in seq_len(length(pl))) {
+    for (j in seq_len(length(pl))) {
+      a1 <- a1 + w[i, j] * h_tr(thetav0pv0[[i]]) * h_tr(thetav0pv0[[j]])
+      a2 <- a2 + w[i, j] * h_tr(thetav0pv0[[i]] %*% thetav0pv0[[j]])
+    }
+  }
+  b <- 1 / (2 * nl) * (a1 + 6 * a2)
+  e <- 1 + a2 / nl
+  e_star <- 1 / (1 - a2 / nl)
+  g <- ((nl + 1) * a1 - (nl + 4) * a2) / ((nl + 2) * a2)
+  denom <- (3 * nl + 2 - 2 * g)
+  c1 <- g / denom
+  c2 <- (nl - g) / denom
+  c3 <- (nl + 2 - g) / denom
+  v_star <- 2 / nl * (1 + c1 * b) / (1 - c2 * b)^2 / (1 - c3 * b)
+  rho <- v_star / (2 * e_star^2)
+  m <- 4 + (nl + 2)  / (nl * rho - 1)
+  lambda <- m / (e_star * (m - 2))
+  list(m = m, lambda = lambda)
+}
+
+#' obtain the adjusted covariance matrix
+#' @param v unadjusted covariance matrix
+#' @param w hessian matrix
+#' @param p P matrix from `h_get_kr_comp`
+#' @param q Q matrix from `h_get_kr_comp`
+#' @param r R matrix from `h_get_kr_comp`
+#' @param linear `logical`\cr whether to use linear Kenward-Roger approximation
+#' @return matrix of adjusted covariance matrix
+#' @keywords internal
+h_var_adj <- function(v, w, p, q, r, linear = FALSE) {
+  if (linear) {
+    r <- matrix(0, nrow = nrow(r), ncol = ncol(r))
+  }
+  dr <- ncol(v)
+  n_theta <- ncol(w)
+  theta_per_group <- nrow(q) / nrow(p)
+  n_groups <- n_theta / theta_per_group
+  ret <- v
```

Comment:
also here a comment to link back to vignette might be helpful. or do that in roxygen chunk already